HBASE-26437 Clean up the znodes for the src after a rename.

joshelser opened the pull request:
HBOSS was orphaning znodes from the src of a path which is renamed. Over
time, this will result in a very large usage of ZK due to HBOSS.

joshelser commented on hbase-oss/src/main/java/org/apache/hadoop/hbase/oss/sync/TreeLockManager.java:
```diff
@@ -488,7 +488,18 @@ public void close() throws IOException {
+            //    /somethingelse  (dst)
+            // Thus, we can't safely delete the znodes for src as it may
+            // then also affect the (held) lock on the dst.
+            recursiveDelete(src);
```
One thing I'm worried about after looking at this (for the third time): is it safe to delete the znodes on the src if the rename was unsuccessful? I need to think about this.

wchevreuil replied:
I think it's ok to delete it. The only con is that anyone trying to access it again would then recreate the znode in writeLock()/readLock().

joshelser commented:
I put both HBASE-26437 and HBASE-26453 into this one PR (as I can't test the latter without the former).

joshelser commented:
`mvn package -Pzk && mvn package -Pzk -Dhadoop.profile=3.2` passed. It looks like `-Plocal` and `-Pnull` may have other issues which need to be corrected.

joshelser commented:
> It looks like -Plocal and -Pnull may have other issues which need to be corrected.

Oh, I'm silly :). `-Pnull` would be expected to fail. `-Plocal` seems to have passed on re-run so maybe a race condition to address later.

I think this is ready to go.

wchevreuil commented on hbase-oss/src/main/java/org/apache/hadoop/hbase/oss/sync/TreeLockManager.java:
```diff
@@ -484,8 +487,29 @@ public AutoLock lockRename(Path rawSrc, Path rawDst) throws IOException {
+            // we only delete the znodes on success.
+            recursiveDelete(src);
+          }
           writeUnlock(src);
```
This call would recreate the znode, as [writeUnlock](https://github.com/apache/hbase-filesystem/blob/f21af3719555a6a0c7c060195f92639690b94423/hbase-oss/src/main/java/org/apache/hadoop/hbase/oss/sync/ZKTreeLockManager.java#L161) calls [get](https://github.com/apache/hbase-filesystem/blob/f21af3719555a6a0c7c060195f92639690b94423/hbase-oss/src/main/java/org/apache/hadoop/hbase/oss/sync/ZKTreeLockManager.java#L368), which in turn  creates the znode. 

I guess lockDelete also has same problem. 

I'm not sure now how curator manages the locks. Is it bound to the specific znode it was created for, so that if the znode got deleted, the lock is gone? If so, we don't need to worry about unlock it after recursiveDelete. Otherwise, we would need to change `ZKTreeLockManager.removeInMemoryLocks` to explicitly release the locks when removing it from the memory cache.

joshelser replied:
> This call would recreate the znode, as writeUnlock calls get, which in turn creates the znode.

Crap.

> I guess lockDelete also has same problem.

Yeah, I had copied what `lockDelete` does in the hopes that it was doing the right thing :)

> I'm not sure now how curator manages the locks. Is it bound to the specific znode it was created for

So, I think the reason that this works is that Curator is creating znodes beneath the .hboss-lock-node. There's a specially formatted znode with READ or WRITE in the name when there is a held lock. I'll have to look into how the heck the unit test is passing...

wchevreuil commented on hbase-oss/src/main/java/org/apache/hadoop/hbase/oss/sync/ZKTreeLockManager.java:
```diff
@@ -241,7 +242,7 @@ protected void recursiveDelete(Path p) throws IOException {
+  synchronized void removeInMemoryLocks(Path p) {
     Iterator<Entry<Path,InterProcessReadWriteLock>> iter = lockCache.entrySet().iterator();
     while (iter.hasNext()) {
       Entry<Path,InterProcessReadWriteLock> entry = iter.next();
```
We may need to explicitly check if we have lock and release it here. See my comment above.

wchevreuil commented on hbase-oss/src/main/java/org/apache/hadoop/hbase/oss/sync/ZKTreeLockManager.java:
```diff
@@ -161,7 +161,7 @@ protected void writeUnlock(Path p) throws IOException {
       get(p).writeLock().release();
```
This 'get' call creates the znode if it's not in the cache, so call to writeUnlock after deleting the znode would create it again.

joshelser replied:
Yeah, right as usual. I want to figure out why my unit test isn't failing like I think it should, given this.

However, I think I can interpret you see this as an improvement, but there is more work to be done (citing the problems, but approving this PR)? Just wanted to double check.

joshelser commented on hbase-oss/src/main/java/org/apache/hadoop/hbase/oss/sync/ZKTreeLockManager.java:
```diff
@@ -252,12 +253,31 @@ private synchronized void removeInMemoryLocks(Path p) {
-  private boolean isBeneath(Path parent, Path other) {
-    if (parent.equals(other)) {
+  /**
+   * Returns true iff the given path is contained beneath the parent path.
```
Oh, this is intentional. Shorthand for "if and only if"

joshelser reviewed (commented):
Just noticed that I also forgot to undo the timeout change so I could step through the tests with a debugger. Let me fix that.

joshelser commented:
Oh, funny. So, my tests do actually clean up ZK, but not for the reason I thought :)

```
2021-11-16 21:20:30,424 [main] DEBUG sync.ZKTreeLockManager (ZKTreeLockManager.java:recursiveDelete(229)) - Removing all mutex and znodes for paths beneath /s3a/embedded/hboss-junit/src
2021-11-16 21:20:30,429 [main] INFO  sync.ZKTreeLockManager (ZKTreeLockManager.java:removeInMemoryLocks(253)) - Not removing /s3a/embedded because it's not beneath /s3a/embedded/hboss-junit/src
2021-11-16 21:20:30,430 [main] INFO  sync.ZKTreeLockManager (ZKTreeLockManager.java:removeInMemoryLocks(253)) - Not removing /s3a/embedded/hboss-junit because it's not beneath /s3a/embedded/hboss-junit/src
2021-11-16 21:20:30,430 [main] INFO  sync.ZKTreeLockManager (ZKTreeLockManager.java:removeInMemoryLocks(253)) - Not removing /s3a because it's not beneath /s3a/embedded/hboss-junit/src
2021-11-16 21:20:30,430 [main] INFO  sync.ZKTreeLockManager (ZKTreeLockManager.java:removeInMemoryLocks(253)) - Not removing /s3a/embedded/hboss-junit/dest because it's not beneath /s3a/embedded/hboss-junit/src
2021-11-16 21:20:30,430 [main] INFO  sync.ZKTreeLockManager (ZKTreeLockManager.java:removeInMemoryLocks(253)) - Not removing /s3a/embedded/hboss-junit/src because it's not beneath /s3a/embedded/hboss-junit/src
2021-11-16 21:20:30,430 [main] INFO  sync.ZKTreeLockManager (ZKTreeLockManager.java:removeInMemoryLocks(253)) - Not removing / because it's not beneath /s3a/embedded/hboss-junit/src
2021-11-16 21:20:30,430 [main] DEBUG sync.ZKTreeLockManager (ZKTreeLockManager.java:writeUnlock(160)) - writeLock /s3a/embedded/hboss-junit/src release
2021-11-16 21:20:30,430 [main] INFO  sync.ZKTreeLockManager (ZKTreeLockManager.java:get(403)) - Lock cached for /s3a/embedded/hboss-junit/src
2021-11-16 21:20:30,434 [main] DEBUG sync.ZKTreeLockManager (ZKTreeLockManager.java:writeUnlock(160)) - writeLock /s3a/embedded/hboss-junit/dest release
2021-11-16 21:20:30,434 [main] INFO  sync.ZKTreeLockManager (ZKTreeLockManager.java:get(403)) - Lock cached for /s3a/embedded/hboss-junit/dest
2021-11-16 21:20:30,439 [main] INFO  oss.TestZNodeCleanup (TestZNodeCleanup.java:teardown(79)) - Dumping contents of ZooKeeper after test from /s3a/embedded/hboss-junit
2021-11-16 21:20:30,439 [main] INFO  oss.TestZNodeCleanup (TestZNodeCleanup.java:printZkBFS(88)) - /s3a/embedded/hboss-junit
2021-11-16 21:20:30,440 [main] INFO  oss.TestZNodeCleanup (TestZNodeCleanup.java:printZkBFS(88)) - /s3a/embedded/hboss-junit/.hboss-lock-znode
2021-11-16 21:20:30,440 [main] INFO  oss.TestZNodeCleanup (TestZNodeCleanup.java:printZkBFS(88)) - /s3a/embedded/hboss-junit/dest
2021-11-16 21:20:30,441 [main] INFO  oss.TestZNodeCleanup (TestZNodeCleanup.java:printZkBFS(88)) - /s3a/embedded/hboss-junit/dest/.hboss-lock-znode
```

This is the logging I added after a successful `mv /src /dest`. We clean up the corresponding node for `/src` in ZK for HBOSS as expected, but I still got `removeInMemoryLocks()` wrong in that I don't remove the Mutex for the parent `/src` znode. That's a bug and the same one I thought I fixed. Go figure.

However, because we don't remove that parent src lock, we can call release on it _without_ recreating the state in ZK. However, we do orphan the Mutex inside Curator/HBOSS which is no good. However, I think this also happens for lockDelete, so I think it might be best to fix that separately. Let me sleep on it and see what I think in the morning.

Thanks for your careful eyes, Wellington. Curious what you think, too.

wchevreuil reviewed (commented):
> Oh, funny. So, my tests do actually clean up ZK, but not for the reason I thought :)
> 
> However, because we don't remove that parent src lock, we can call release on it _without_ recreating the state in ZK. However, we do orphan the Mutex inside Curator/HBOSS which is no good.

Yep, just had debugged your `TestZNodeCleanup.testRename`, could confirm the znode cleanup is working, but that's because `removeInMemoryLocks` is not removing the path from `ZKTreeLockManager` cache. 

> However, I think this also happens for lockDelete, so I think it might be best to fix that separately. Let me sleep on it and see what I think in the morning.

Could do the `removeInMemoryLocks` separately in HBASE-26453, but I think then znodes would get created again when unlocking, failing this PR tests. So, once we fix `removeInMemoryLocks`, we need to make sure `rename` and `delete` woild not recreate the path again when calling `unlock`.
